### PR TITLE
Fix some assertions in debug mode

### PR DESCRIPTION
**Pull request by dsyme**
WHile fixing #2793 I improved the diagnostics for some assertions and remove some exceptions that were happening on the startup path

I also cleaned up the logic for ``BinFolderOfDefaultFSharpCompiler`` a little and take advantage of the fact that optdata/sigdata are gone to simplify some things.

### CODE REVIEW

**Comment by dsyme** [comment]
Also includes @saul's fix for https://github.com/Microsoft/visualfsharp/issues/3005

File: vsintegration/src/FSharp.Editor/CodeFix/SimplifyName.fs [hunk 1]
```diff
@@ -13,7 +13,7 @@ open Microsoft.CodeAnalysis.Text
 open Microsoft.CodeAnalysis.CodeFixes
 open Microsoft.CodeAnalysis.CodeActions
 
-[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = PredefinedCodeFixProviderNames.SimplifyNames); Shared>]
+//[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = PredefinedCodeFixProviderNames.SimplifyNames); Shared>]
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/CodeFix/SimplifyName.fs]: I assume you didn't mean to disable this? There's an option for this in the options dialog.

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/CodeFix/SimplifyName.fs]: 👍

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/CodeFix/SimplifyName.fs]: Thanks :)  I'll re-enable it.  However see also https://github.com/Microsoft/visualfsharp/issues/3009

**Pull request merged**